pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
+
+    @property
+    def active(self):
+        """Check if an unlocking context manager is active."""
+        return self._lock_id is not None
+
+    def lock(self):
+        """Lock the object."""
+        self._locked = True
+        self._lock_id = None
+
+    def unlock(self, manager_id=None):
+        """Unlock the object."""
+        self._locked = False
+        if manager_id is not None:
+            self._lock_id = manager_id
+
+
+class UnlockInstance:
+    """Context manager that temporarily unlocks an immutable instance
+    of ``CCLObject``.
+
+    Parameters:
+        instance (``CCLObject``):
+            Instance of ``CCLObject`` to unlock within the scope
+            of the context manager.
+        mutate (``bool``):
+            If the enclosed function mutates the object, the stored
+            representation is automatically deleted.
+    """
+
+    def __init__(self, instance, mutate=True):
+        self.instance = instance
+        self.mutate = mutate
+        # Define these attributes for easy access.
+        self.id = id(self)
+        self.thread_lock = RLock()
+        # We want to catch and exit if the instance is not a CCLObject.
+        # Hopefully this will be caught downstream.
+        self.check_instance = isinstance(instance, CCLObject)
+        if self.check_instance:
+            self.object_lock = instance._object_lock
+
+    def __enter__(self):
+        if not self.check_instance:
+            return
+
+        with self.thread_lock:
+            # Prevent simultaneous enclosing of a single instance.
+            if self.object_lock.active:
+                # Context manager already active.
+                return
+
+            # Unlock and store the fingerprint of this context manager so that
+            # only this context manager is allowed to run on the instance.
+            self.object_lock.unlock(manager_id=self.id)
+
+    def __exit__(self, type, value, traceback):
+        if not self.check_instance:
+            return
+
+        # If another context manager is running,
+        # do nothing; otherwise reset.
+        if self.id != self.object_lock._lock_id:
+            return
+
+        with self.thread_lock:
+            # Reset `repr` if the object has been mutated.
+            if self.mutate:
+                try:
+                    delattr(self.instance, "_repr")
+                    delattr(self.instance, "_hash")
+                except AttributeError:
+                    # Object mutated but none of these exist.
+                    pass
+
+            # Lock the instance on exit.
+            self.object_lock.lock()
+
+
+def unlock_instance(func=None, *, argv=0, mutate=True):
+    """Decorator that temporarily unlocks an instance of CCLObject.
+
+    Arguments:
+        func (``function``):
+            Function which changes one of its ``CCLObject`` arguments.
+        argv (``int``):
+            Which argument should be unlocked. Defaults to the first argument.
```

Comment:
Took me a while. Can you specify here that said argument is meant to be a CCLObject?

Reply:
Also, can we only use this for one CCLObject at a time? E.g. I could imagine a function that modifies both a tracer and a profile.

Reply:
There's the context manager `UnlockInstance` which does the bulk of the work, and then there's the helper function `unlock_instance` which can be used as a decorator. Currently, `unlock_instance` is only meant to be used as a function decorator, and is agnostic to the variable name of what is being mutated. Implicitly, this is assumed to be the first argument of the function (`self`), because the only reason I created this one was to use it inside class methods which mutate their objects.

If multiple `CCLObject`s need to be mutated at a time, this can simply be done using stacked context managers. In theory, I could extend the functionality to accept multiple objects, but I think it would unnecessarily complicate the code, especially as `mutate` needs to be specified for all of them.

Let me know what you think.

Reply:
Imagine for example the following scenario:
```python
cosmo0 = ccl.CosmologyVanillaLCDM()
cosmo1 = ccl.CosmologyVanillaLCDM()
```
The following will work
```python
# Nested context managers for multiple attribute assignments.
with ccl.UnlockInstance(cosmo0):
    with ccl.UnlockInstance(cosmo1):
        cosmo0.test = 0
        cosmo1.test = 1
```
and the following (neater) way will work as well
```python
# Can be sequenced like so
with ccl.UnlockInstance(cosmo0), ccl.UnlockInstance(cosmo1):
    cosmo0.test = 2
    cosmo1.test = 3
```

If what you are proposing is something like
```python
with ccl.UnlockInstance(cosmo0, cosmo1, mutate=[True, False]):
    cosmo0.test = 2
    cosmo1.test = 3
```
I don't think there's much benefit to it.

Reply:
ok, specify that args[argv] is meant to be a CCLObject in the comment and we're done with this

Reply:
I have, and I have pushed it. Not sure why it is not showing.

Reply:
It is showing now.